twilight-gateway-queue/src/in_memory.rs

```diff
@@ -0,0 +1,272 @@
+                        // Shards *must* identify in order.
+                        yield_now().await;
+                        remaining -= 1;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Memory based [`Queue`] implementation backed by an efficient background task.
+///
+/// [`InMemoryQueue::update`] allows for dynamically changing the queue's
+/// settings.
+///
+/// Cloning the queue is cheap and just increments a reference counter.
+///
+/// # Settings
+///
+/// `remaining` is reset to `total` after `reset_after` and then every
+/// [`LIMIT_PERIOD`].
+///
+/// A `max_concurrency` of `0` processes all requests instantly, effectively
+/// disabling the queue.
+#[derive(Clone, Debug)]
+pub struct InMemoryQueue {
+    /// Sender to communicate with the background [task runner].
+    ///
+    /// [task runner]: runner
+    tx: mpsc::UnboundedSender<Message>,
+}
+
+impl InMemoryQueue {
+    /// Creates a new `InMemoryQueue` with custom settings.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `total` < `remaining`.
+    pub fn new(max_concurrency: u8, remaining: u16, reset_after: Duration, total: u16) -> Self {
+        assert!(total >= remaining);
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        tokio::spawn(runner(
+            rx,
+            Settings {
+                max_concurrency,
+                remaining,
+                reset_at: Instant::now() + reset_after,
+                total,
+            },
+        ));
+
+        Self { tx }
+    }
+
+    /// Update the queue with new info from the [Get Gateway Bot] endpoint.
+    ///
+    /// May be regularly called as the bot joins/leaves guilds.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use twilight_gateway_queue::InMemoryQueue;
+    /// # struct GetGatewayBot {
+    /// #     session_start_limit: SessionStartLimit,
+    /// # }
+    /// # struct SessionStartLimit {
+    /// #     max_concurrency: u8,
+    /// #     remaining: u16,
+    /// #     reset_after: u64,
+    /// #     total: u16,
+    /// # }
+    /// # let rt = tokio::runtime::Builder::new_current_thread()
+    /// #     .enable_time()
+    /// #     .build()
+    /// #     .unwrap();
+    /// use std::time::Duration;
+    ///
+    /// async fn get_gateway_bot() -> GetGatewayBot {
+    ///     unimplemented!()
+    /// }
+    ///
+    /// # rt.block_on(async {
+    /// # let queue = InMemoryQueue::default();
+    /// let session = get_gateway_bot().await.session_start_limit;
+    /// queue.update(
+    ///     session.max_concurrency,
+    ///     session.remaining,
+    ///     Duration::from_millis(session.reset_after),
+    ///     session.total,
+    /// );
+    /// # })
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if `total` < `remaining`.
+    ///
+    /// [Get Gateway Bot]: https://discord.com/developers/docs/topics/gateway#get-gateway-bot
```

Comment:
This example just looks super awkward because you're not pulling in model or http as dev-dependencies, why not just do that?